### DISCUSSION TITLE
Format on Modifications

### DISCUSSION THREAD

**Discussion by Jiyangson** [LazyVim/LazyVim#4150] [Q&A] [answered]
I think the formatter that Lazyvim use is Conform.nvim right? I would like to set the formatter to format on modifications, but I didn't find this feature in Conform. I found this plugin https://github.com/joechrisellis/lsp-format-modifications.nvim, and followed it's instruction: 
![image](https://github.com/user-attachments/assets/7eb5ea3e-9e6f-4d3b-997c-20fe3d43f3a6)

This will report an error when I format: 
![image](https://github.com/user-attachments/assets/28528685-8b93-462c-bf45-8020f3d57d06)

How to properly setup format on modifications? Thanks a lot.

**Comment by rogeriofrsouza** [comment]
The reason is [in the docs](https://github.com/joechrisellis/lsp-format-modifications.nvim?tab=readme-ov-file#requirements)

> Any LSP server that you want to use with this plugin must support the [DocumentRangeFormattingProvider server capability](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.languageserver.protocol.servercapabilities.documentrangeformattingprovider?view=visualstudiosdk-2022) — lsp-format-modifications.nvim will warn if an unsupported LSP server is used.

Idk much about the Copilot LSP, can you check if it has this capability?

**Reply by Jiyangson** [reply]
I think the problem should be related to my config, because I was formatting a python file and using ruff as the lsp server. Also I open copilot by default. I wish the plugin to identify the ruff lsp which I assume support the Document.

**Reply by Jiyangson** [reply]
nevermind, I disabled copilot and pyright(which does not support the Document), only using ruff, and  It still could not format python file. I tried it with cpp file using clangd and it worked. It seems to be a python lsp server problem. Thanks a lot.

**Comment by Jiyangson** [comment, marked as answer]
A year later I finally have some time to debug it. And I would like to share a way that works for me.
Firstly, there are multiple LSP servers activated in python, e.g. pyright, ruff, copilot, and as the picture shows in the description, servers like copilot or pyright (maybe?) do not have the range formatting behavior. So I just hack the client to be ruff.
 Secondly, default lazyvim autocmd for formatting must be disabled otherwise it may conflict with format_modification. That is set `vim.g.autoformat = false` in options.lua
This is the final autocmd that I use.
```{lua}
vim.api.nvim_create_autocmd("LspAttach", {
  group = vim.api.nvim_create_augroup("lsp_attach_auto_diag", { clear = true }),
  callback = function(args)
    -- the buffer where the lsp attached
    ---@type number
    local buffer = args.buf
    local ruff_client = nil
    for _, client in pairs(vim.lsp.get_clients({ bufnr = buffer })) do
      if client.name == "ruff" then
        ruff_client = client
        break
      end
    end
    if not ruff_client then
      return
    end

    local augroup_id = vim.api.nvim_create_augroup("FormatModificationsDocumentFormattingGroup", { clear = false })
    vim.api.nvim_clear_autocmds({ group = augroup_id, buffer = buffer })

    vim.api.nvim_create_autocmd({ "BufWritePre" }, {
      group = augroup_id,
      buffer = buffer,
      callback = function()
        local lsp_format_modifications = require("lsp-format-modifications")
        lsp_format_modifications.format_modifications(ruff_client, buffer)
      end,
    })
  end,
})
```
